---
title: "VCF validation - With Hap.py and GIAB data"
author: "Clalit Genomics"
date: "11/8/2021"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
    fig_width: 12
    fig_height: 8
---

```{r variable_setup,eval=F,echo=F}
# define analysis folder 
# if you want to re-analyze a previous folder, just change the name to the appropriate folder
library(glue)
analysis_name<-glue('giab_202206_{Sys.Date()}')
analysis_dir<-glue('{analysis_name}')
if (!dir.exists(glue('./output/{analysis_dir}'))){dir.create(glue('./output/{analysis_dir}'))}

analysis_build<-'hg19'

#target_region<-'/media/SSD/Bioinformatics/Databases/idt/xgen-exome-research-panel-v2-targets-hg19.bed'
#target_region<-'/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/data/stratifications/refseq_hg19_curated_cds_20220414.bed.gz'
target_region<-NA  # NA - means that there is no target region - run the analysis on the whole genome

vcf_stats=T # whether or not to perform vcf stats collection
```


```{r render_report,eval=F,echo=F}
library(rmarkdown)
output_dir <- glue("/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/output/{analysis_dir}/report")
render('/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/src/sequencing_validation_analysis.Rmd', output_dir = output_dir,output_file = sprintf('vcf_validation_report.%s.html',Sys.Date()))
```

```{r var_setup, include=TRUE,echo=1:3}

happy_path<-'/media/SSD/Bioinformatics/Tools/hap.py-build/bin/hap.py'
reference<-ifelse(analysis_build=='hg19',
                  '/media/SSD/Bioinformatics/Databases/hg19/hg19.fa',
                  '/media/SSD/Bioinformatics/Databases/hg38/hg38.fa')

vcfeval_template<-ifelse(analysis_build=='hg19',
                     '/media/SSD/Bioinformatics/Databases/hg19/hg19.SDF',
                     '/media/SSD/Bioinformatics/Databases/hg38/hg38_SDF')


```

```{r markdown_setup,include=FALSE}
project_dir<-'/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()
```

```{r run_happy,eval=TRUE,echo=FALSE,results=FALSE}
# validate the target region
if (!is.na(target_region)){
  #validate_input_bed_file(target_region)
  target_region_fixed<-prep_target_file_for_happy(target_region,output_dir = glue('/media/SSD/Bioinformatics/Projects/sequencing_validation/sequencing_validation_project/output/{analysis_dir}'))
}

for (i in 1:nrow(samples_sheet)) {
#for (i in 1:1) {
  sample <- samples_sheet %>% slice(i)
  giab_ref <- giab_sheet %>% filter(build==analysis_build & name == sample$truth)
  truth <- giab_ref %>% pull(vcf)
  fp_bed <- giab_ref %>% pull(false_positive)
  query <- sample %>% pull(query_path)
  output_prefix <- sample %>% pull(query_name)
  if (is_sample_already_analyzed(glue('{analysis_dir}/{output_prefix}'))) {
    next
  }# make sure sample was not already ran
  info(logger, print(
    sprintf(
      'Analyzing %s, comparing against %s..',
      output_prefix,
      giab_ref$name
    )
  ))
  #message(sprintf('Analyzing %s, comparing against %s..',output_prefix,giab_ref$name))
  run_happy(
    happy_path,
    truth,
    query,
    reference = reference,
    target_region = target_region_fixed,
    fp_bed = fp_bed,
    vcfeval_template = vcfeval_template,
    output_prefix = output_prefix,
    analysis_dir = analysis_dir
  )
  info(logger, print(sprintf('Done analyzing %s.', output_prefix)))
}

```

# Sample sheet

```{r sample_sheet_table}
DT::datatable(samples_sheet,
              options=list(scrollX=T),
              filter = list(position = 'top', clear = FALSE))
```

# Stratifications 

```{r stratifications_stats}

DT::datatable(strat_files_stats,
              options=list(scrollX=T),
              filter = list(position = 'top', clear = FALSE))
```

# VCF stats

### VCFs stats table

```{r vcf_stats,out.width="150%",include=TRUE,message=FALSE,warnings=FALSE}
all_vcf_stats<-get_all_vcf_stats_with_stratifications()
if (vcf_stats){
  info(logger,print(sprintf('Producing VCF stats..')))
  all_vcf_stats<-get_all_vcf_stats_with_stratifications()
  DT::datatable(all_vcf_stats%>%select(-Location),
                options=list(scrollX=T),
                filter = list(position = 'top', clear = FALSE))
}

```

### VCF stats plots {.tabset}

```{r, results = 'asis',fig.width=12,fig.height=8}
for (target in unique(all_vcf_stats$target)){
  cat("#### ", target, "\n")
  p<-plot_vcfstats_results(all_vcf_stats,target_filter=target)
  #cat(knitr::knit_print(print(p)))
  knitr::knit_print(p)
  cat("\n\n")
}
```

# Hap.py results: All samples

```{r happy_results_table,eval=T}
# retrieve the happy outcome files
happy_results<-collect_happy_results(samples_sheet,analysis_dir)
happy_results<-happy_results%>%mutate(across(where(is.character),as.factor))
DT::datatable(happy_results,
              extensions = 'Buttons', 
              options=list(scrollX=T,dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')),
              filter = list(position = 'top', clear = FALSE))
```

# Hap.py results: Per Sample {.tabset}

```{r happy_results_per_sample,eval=T, results='asis',warning=F}
# retrieve the happy outcome files
happy_results<-collect_happy_results(samples_sheet,analysis_dir)
happy_results<-happy_results%>%mutate(across(where(is.character),as.factor))
for (sample_name in unique(happy_results$query_name)){
  cat('## ',sample_name,'\n')
  sample_table<-generate_sample_table(happy_results,sample_name)
  print(tagList(
    DT::datatable(
      sample_table,
      extensions = 'Buttons',
      options = list(
        scrollX = T,
        dom = 'Bfrtip',
        buttons = c('copy', 'csv', 'excel')
      ),
      filter = list(position = 'top', clear = FALSE)
    ),
    br(),
    br()
  ))
  cat('\n\n')
}
cat('\n\n')
```

# Hap.py results: Aggregated

```{r happy_summary_table,eval=T,warning=F,message=F}

# summarize results
happy_summary<-happy_summary_per_group(happy_results)
happy_summary<-happy_summary%>%mutate(across(where(is.character),as.factor))
DT::datatable(
  happy_summary,
  extensions = 'Buttons',
  options = list(
    scrollX = T,
    dom = 'Bfrtip',
    buttons = c('copy', 'csv', 'excel')
  ),
  filter = list(position = 'top', clear = FALSE)
)

```


# Hap.py results: Plots {.tabset}

```{r happy_plots, results='asis',eval=T,out.width='150%'}
# Plot the results
for (subset in unique(happy_results$Subset)){
  if (subset=='TS_boundary'){next}
  cat('## Subset:',subset,'\n')
  plot_happy_results(happy_results,
             query_group_name=NA,
             subset = c(subset),
             subtype = c('*'),
             filter = c('PASS'),
             var_type = c('SNP','INDEL'),
             facet_by=c())
  cat('\n\n')
}

for (subtype in unique(happy_results$Subtype)){
  if (grepl('^C',subtype)){next}
  cat('## Subtype:',subtype,'\n')
  plot_happy_results(happy_results,
             query_group_name=NA,
             subset = c('*'),
             subtype = c(subtype),
             filter = c('PASS'),
             var_type = c('SNP','INDEL'),
             facet_by=c())
  cat('\n\n')
}
```